lib/engine/game/g_1873/game.rb

```diff
@@ -683,40 +839,148 @@ def pres_change_ok?(corporation)
           public_mine?(corporation) || corporation.operated?
         end
 
-        def get_machine(mine)
-          mine.trains.find { |t| t.name.include? 'M' }
+        def machine(mine)
+          mine.trains.find { |t| train_is_machine?(t) }
         end
 
-        def get_switcher(mine)
-          mine.trains.find { |t| t.name.include? 'S' }
+        def machine_size(mine)
+          machine(mine)&.distance || 1
         end
 
-        def machine_size(mine)
-          get_machine(mine)&.distance || 1
+        def switcher(mine)
+          mine.trains.find { |t| train_is_switcher?(t) }
         end
 
         def switcher_size(mine)
-          get_switcher(mine)&.distance
+          switcher(mine)&.distance
+        end
+
+        def mhe_income
+          @mhe.trains.first.distance * 100
+        end
+
+        def public_mine_slots(entity)
+          return 0 unless public_mine?(entity)
+
+          @corporation_info[entity][:slots]
+        end
+
+        def public_mine_mines(entity)
+          return [] unless public_mine?(entity)
+
+          @corporation_info[entity][:mines]
         end
 
-        def maintenance_costs(_corp)
-          0
+        # find the best mine slot for a machine out of the legal candidates (ones that have a smaller machine)
+        def find_machine_slot(entity, new_size)
+          mines = public_mine_mines(entity)
+          candidates = mines.each_index.map { |i| i }.select { |i| machine_size(mines[i]) < new_size }
+          candidates.max do |c|
+            mines.each_index.map { |i| i }.sum do |i|
+              calculate_mine_revenue(mines[i], (c == i ? new_size : machine_size(mines[i])), nil)
+            end
+          end
+        end
+
+        # find the best mine slot for a switcher, first based on switcher sizes, then maximizing
+        # revenue
+        def find_switcher_slot(entity, new_size)
+          mines = public_mine_mines(entity)
+          min_size = mines.map { |m| switcher_size(m) || 0 }.min
+          candidates = mines.each_index.map { |i| i }.select { |i| (switcher_size(mines[i]) || 0) == min_size }
```

Comment:
why are you doing index  map identity?

Reply:
I want an array of indices.

This method is going away.

Reply:
you would do mines.select.with_index instead

Reply:
Yeah, I missed that you could use with_index on more methods than each. That explains why my code is so ugly.

Reply:
method deleted